[ Dy2static ] select input fix and while_op memory bug fixed.

2742195759 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
Others
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
Others
### Describe
<!-- Describe what this PR does -->
select input fix  + while_loop risky assign warning

2742195759 committed:
while support for python container.
It is convenient to convert more dynamic graph codes into static graphs.

2742195759 committed:
cond support python container

2742195759 committed:
1. make select_input output shape = input[1]
2. add warning in while_loop risky assign

2742195759 committed:
Merge remote-tracking branch 'upstream/develop' into dy2static-select-input

Aurelius84 commented on python/paddle/fluid/layers/control_flow.py:
```diff
@@ -1190,6 +1190,13 @@ def assign_skip_lod_tensor_array(input, output):
     Assign input to output, but skip the process of copying LoDTensorArray unless it's created in while_block.
     """
+
+    def _shape_diff(vara, varb):
```
```suggestion
    def  has_shape_diff(x, y):
```

Aurelius84 commented on python/paddle/fluid/layers/control_flow.py:
```diff
@@ -1190,6 +1190,13 @@ def assign_skip_lod_tensor_array(input, output):
+
+    def _shape_diff(vara, varb):
+        if len(vara.shape) != len(varb.shape): return True
+        for sa, sb in zip(vara.shape, varb.shape):
```
```suggestion
        for a, b in zip(x.shape, y.shape):
```

2742195759 committed:
fix 2 problem in GPT export:
1. a bug in while_op no_need_copy_var, which causes gpu memory leakage
2. a bug in undefined_var where the stop_gradient should be False.

2742195759 committed:
change name by code review

2742195759 committed:
format

Aurelius84 reviewed (approved):
LGTM
